### PR TITLE
Adding artifact assembler tool which consolidates the artifacts and produce the package

### PR DESCRIPTION
**Pull request by kshyju**
Adding artifact assembler program which collects artifacts from different sources and assemble the final package for core tools.

Current version gets inproc6,inproc8, custom host artifacts and assemble a package where inproc6 and inproc8 are child folders in the package and custom host executable will be in the root.

This program will generate the consolidated artifact zip. The content of the zip will look like the following.

image

The "in-proc6" and "in-proc8" folders are core tools artifacts produced for net6 and net8 target frameworks.

### Pull request checklist

* [x] My changes **do not** require documentation changes
  * [ ] Otherwise: Documentation issue linked to PR
* [x] My changes **do not** need to be backported to a previous version
  * [ ] Otherwise: Backport tracked by issue/PR #issue_or_pr
* [ ] I have added all required tests (Unit tests, E2E tests)

### CODE REVIEW

File: src/Azure.Functions.ArtifactAssembler/Program.cs [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+{
+    internal class Program
+    {
+        static async Task<int> Main(string[] args)
+        {
+            try
+            {
+                var currentWorkingDirectory = Environment.CurrentDirectory;
+
+                // print child directories present in currentWorkingDirectory.
```
Comment (kshyju) [inline src/Azure.Functions.ArtifactAssembler/Program.cs:14]: May clean this up (hide behind a verbose variable /flag) in a future version.

File: src/Azure.Functions.ArtifactAssembler/ArtifactAssembler.cs [hunk 2]
```diff
@@ -0,0 +1,216 @@
...
+
+                // Copy in-proc8 files
+                await Task.Run(() => FileUtilities.CopyDirectory(runtimeDir, Path.Combine(targetRuntimeDir, InProc8DirectoryName)));
+
+                // Copy in-proc6 files if exist
+                var inProc6Dir = Path.Combine(_inProc6ExtractedRootDir, targetRuntimeName);
+                if (Directory.Exists(inProc6Dir))
+                {
+                    await Task.Run(() => FileUtilities.CopyDirectory(inProc6Dir, Path.Combine(targetRuntimeDir, InProc6DirectoryName)));
+                }
```
Comment (mattchenderson) [inline src/Azure.Functions.ArtifactAssembler/ArtifactAssembler.cs:153]: This feels surprising different from the .NET 8 version above. Should .NET 6 not also throw, for example? Would it make sense to encapsulate this logic and then just invoke it for both? The same question applies to a few places earlier on as well.

File: src/Azure.Functions.ArtifactAssembler/ArtifactAssembler.cs [hunk 3]
```diff
@@ -0,0 +1,216 @@
...
+        }
+
+        private string GetRIdForArtifactName(string artifactName)
+        {
+            if (_artifactNamesToToRIdMap.TryGetValue(artifactName, out var rid))
+            {
+                return rid;
+            }
+
+            throw new InvalidOperationException($"rid not found for artifact name {artifactName}");
```
Comment (mattchenderson) [inline src/Azure.Functions.ArtifactAssembler/ArtifactAssembler.cs:191]: Nit: acronym should be all caps, but honestly we should expand the acronym fully in general.

**Review by mattchenderson — COMMENTED** [review]
General questions:

- When is this built and run? How does it interface with the CI pipelines?
- What are the instructions for running this in a local context, if appropriate?
- Conceptually, does this actually belong in the `/src` folder? It feels more connected to the build and CI processes.

File: src/Azure.Functions.ArtifactAssembler/ArtifactAssembler.cs [hunk 4]
```diff
@@ -0,0 +1,216 @@
...
+    internal sealed class ArtifactAssembler
+    {
+        private const string StagingDirName = "staging";
+        private const string InProc8DirectoryName = "in-proc8";
+        private const string InProc6DirectoryName = "in-proc6";
+
+        /// <summary>
+        /// The artifact names for which we want to pack a custom host with it.
+        /// </summary>
+        private readonly string[] _customHostTargetArtifactZipNames =
```
[the remark this answers is not in the document]
Comment (kshyju) [reply src/Azure.Functions.ArtifactAssembler/ArtifactAssembler.cs]: Yes, simplified.

File: src/Azure.Functions.ArtifactAssembler/ArtifactAssembler.cs [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+            { "Azure.Functions.Cli.min.win-arm64", "win-arm64" }
+        };
+
+        private readonly string _inProcArtifactDirectoryName = string.Empty;
+        private readonly string _coreToolsHostArtifactDirectoryName = string.Empty;
+        private readonly string _inProc6ArtifactZipName = string.Empty;
+        private readonly string _inProc8ArtifactZipName = string.Empty;
+        private readonly string _coreToolsHostArtifactZipName = string.Empty;
+        private readonly string _rootWorkingDirectory;
+        private readonly string _stagingDirectory;
```
[the remark this answers is not in the document]
Comment (kshyju) [reply src/Azure.Functions.ArtifactAssembler/ArtifactAssembler.cs:28]: Fixed the others to be consistent.

File: src/Azure.Functions.ArtifactAssembler/EnvironmentVariables.cs [hunk 6]
```diff
@@ -0,0 +1,15 @@
...
+
+namespace Azure.Functions.ArtifactAssembler
+{
+    internal class EnvironmentVariables
+    {
+        internal static readonly string InProcArtifactAlias = "IN_PROC_ARTIFACT_ALIAS";
+        internal static readonly string CoreToolsHostArtifactAlias = "CORETOOLS_HOST_ARTIFACT_ALIAS";
+        internal static readonly string InProc6ArtifactZipName = "IN_PROC6_ARTIFACT_ZIP_NAME";
+        internal static readonly string InProc8ArtifactZipName = "IN_PROC8_ARTIFACT_ZIP_NAME";
+        internal static readonly string CoreToolsHostArtifactZipName = "CORETOOLS_HOST_ARTIFACT_ZIP_NAME";
```
[the remark this answers is not in the document]
Comment (kshyju) [reply src/Azure.Functions.ArtifactAssembler/EnvironmentVariables.cs]: Yes. They should be. Fixed.

**Review by aishwaryabh — COMMENTED** [review]
can you attach a screenshot of what the final product looks locally like after running the tool? it would be great to have a visual!

File: src/Azure.Functions.ArtifactAssembler/ArtifactAssembler.cs [hunk 7]
```diff
@@ -0,0 +1,214 @@
...
+        private const string InProc6DirectoryName = "in-proc6";
+
+        /// <summary>
+        /// The artifacts for which we want to pack a custom host with it.
+        /// This dictionary contains the artifact name and the corresponding runtime identifier value.
+        /// </summary>
+        private readonly Dictionary<string, string> _customHostArtifacts = new()
+        {
+            { "Azure.Functions.Cli.min.win-x64", "win-x64" },
+    //        { "Azure.Functions.Cli.min.win-arm64", "win-arm64" }
```
Comment (aishwaryabh) [inline src/Azure.Functions.ArtifactAssembler/ArtifactAssembler.cs:19]: should this be commented out?

Comment (kshyju) [reply src/Azure.Functions.ArtifactAssembler/ArtifactAssembler.cs]: Removed. Commented out for local testing

File: src/Azure.Functions.ArtifactAssembler/ArtifactAssembler.cs (hunk 3 above)
Comment (kshyju) [reply src/Azure.Functions.ArtifactAssembler/ArtifactAssembler.cs]: Fixed.

File: src/Azure.Functions.ArtifactAssembler/ArtifactAssembler.cs (hunk 2 above)
Comment (kshyju) [reply src/Azure.Functions.ArtifactAssembler/ArtifactAssembler.cs]: Yes, both should throw. I fixed it now.

File: src/Azure.Functions.ArtifactAssembler/ArtifactAssembler.cs [hunk 8]
```diff
@@ -0,0 +1,216 @@
...
+        private string _inProc8ExtractedRootDir = string.Empty;
+        private string _coreToolsHostExtractedRootDir = string.Empty;
+                
+        internal ArtifactAssembler(string rootWorkingDirectory)
+        {
+            _inProcArtifactDirectoryName = GetRequiredEnvironmentVariable(EnvironmentVariables.InProcArtifactAlias);
+            _coreToolsHostArtifactDirectoryName = GetRequiredEnvironmentVariable(EnvironmentVariables.CoreToolsHostArtifactAlias);
+            _inProc6ArtifactZipName = GetRequiredEnvironmentVariable(EnvironmentVariables.InProc6ArtifactZipName);
+            _inProc8ArtifactZipName = GetRequiredEnvironmentVariable(EnvironmentVariables.InProc8ArtifactZipName);
+            _coreToolsHostArtifactZipName = GetRequiredEnvironmentVariable(EnvironmentVariables.CoreToolsHostArtifactZipName);
```
[the remark this answers is not in the document]
Comment (kshyju) [reply src/Azure.Functions.ArtifactAssembler/ArtifactAssembler.cs]: discussed offline. But command line support can be added later.

**Comment by kshyju** [comment]
> can you attach a screenshot of what the final product looks locally like after running the tool? it would be great to have a visual!

It's in the PR description now.

File: src/Azure.Functions.ArtifactAssembler/REAME.MD [hunk 9]
```diff
@@ -0,0 +1,49 @@
+# Azure Functions Artifact Assembler
+
+This project is designed to assemble artifacts for Azure Functions, including custom core tools hosts and various versions of in-proc artifacts (to support .NET 6.0 and .NET 8.0 in-proc apps). It extracts, processes, and packages these artifacts into a format suitable for use with Visual Studio.
```
Comment (aishwaryabh) [inline src/Azure.Functions.ArtifactAssembler/REAME.MD:3]: should we specify that we currently package VS builds with this, but we will expand to the other use cases; just to not confuse others for the full context

File: src/Azure.Functions.ArtifactAssembler/REAME.MD [hunk 10]
```diff
@@ -0,0 +1,49 @@
...
+
+- .NET 8 SDK
+- Visual Studio 2022 or later
+
+## Expected Artifact Folders
+
+The following artifact folders are expected to be present in the root directory of the executable (when debugging locally, it will be `bin/debug/net8.0`):
+
+- A directory which contains `in-proc6` and `in-proc8` artifact directories (not zips).
+- A directory which contains the core tools host artifact directory (not zip).
```
Comment (aishwaryabh) [inline src/Azure.Functions.ArtifactAssembler/REAME.MD:15]: maybe specify that the core tools artifact directory is the root which contains `inproc6` and `inproc8` directories

**Review by aishwaryabh — APPROVED** [review]

File: src/Azure.Functions.ArtifactAssembler/REAME.MD (hunk 9 above)
Comment (kshyju) [reply src/Azure.Functions.ArtifactAssembler/REAME.MD:3]: Updated the language.

File: src/Azure.Functions.ArtifactAssembler/REAME.MD (hunk 10 above)
Comment (kshyju) [reply src/Azure.Functions.ArtifactAssembler/REAME.MD:15]: core tools host artifact directory is not the root which contains inproc6 and inproc8. core tools host artifact directory is the one where custom host artifact directory is present.

**Pull request merged**